src/ripple/app/tx/impl/SetCron.cpp

```diff
@@ -0,0 +1,274 @@
+#include <ripple/protocol/st.h>
+
+namespace ripple {
+
+TxConsequences
+SetCron::makeTxConsequences(PreflightContext const& ctx)
+{
+    return TxConsequences{ctx.tx, TxConsequences::normal};
+}
+
+NotTEC
+SetCron::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureCron))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+    auto& j = ctx.j;
+
+    if (tx.getFlags() & tfCronSetMask)
+    {
+        JLOG(j.warn()) << "SetCron: Invalid flags set.";
+        return temINVALID_FLAG;
+    }
+
+    // DelaySeconds (D), RepeatCount (R)
+    // DR - Set Cron with Delay and Repeat
+    // D- - Set Cron (once off) with Delay only (repat implicitly 0)
+    // -R - Invalid
+    // -- - Clear any existing cron (succeeds even if there isn't one) / with
+    // tfCronUnset flag set
+
+    bool const hasDelay = tx.isFieldPresent(sfDelaySeconds);
+    bool const hasRepeat = tx.isFieldPresent(sfRepeatCount);
+
+    if (hasRepeat && !hasDelay)
+    {
+        JLOG(j.warn()) << "SetCron: DelaySeconds must also be specified when "
+                          "RepeatCount is present.";
+        return temMALFORMED;
+    }
+
+    if (hasDelay)
+    {
+        auto delay = tx.getFieldU32(sfDelaySeconds);
+        if (delay > 31536000UL /* 365 days in seconds */)
+        {
+            JLOG(j.debug()) << "SetCron: DelaySeconds was too high. (max 365 "
+                               "days in seconds).";
+            return temMALFORMED;
+        }
+    }
+
+    if (hasRepeat)
+    {
+        auto recur = tx.getFieldU32(sfRepeatCount);
+        if (recur > 256)
+        {
+            JLOG(j.debug())
+                << "SetCron: RepeatCount too high. Limit is 256. Issue "
+                   "new SetCron to increase.";
+            return temMALFORMED;
+        }
+    }
+
+    if (tx.isFlag(tfCronUnset))
+    {
+        if (hasDelay || hasRepeat)
+        {
+            JLOG(j.warn()) << "SetCron: tfCronUnset flag cannot be used with "
+                              "DelaySeconds or RepeatCount.";
+            return temMALFORMED;
+        }
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+SetCron::preclaim(PreclaimContext const& ctx)
+{
+    if (ctx.tx.isFlag(tfCronUnset))
+    {
+        auto const account = ctx.tx.getAccountID(sfAccount);
+        auto const sle = ctx.view.read(keylet::account(account));
+        if (!sle)
+            return tefINTERNAL;
+
+        if (!sle->isFieldPresent(sfCron))
+            return tecNO_ENTRY;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+SetCron::doApply()
+{
+    auto& view = ctx_.view();
+    auto const& tx = ctx_.tx;
+
+    bool const isDelete = tx.isFlag(tfCronUnset);
+
+    // delay can be zero, in which case the cron will usually execute next
+    // ledger.
+    uint32_t delay{0};
+    uint32_t recur{0};
+
+    if (!isDelete)
+    {
+        delay = tx.getFieldU32(sfDelaySeconds);
```

Comment:
wouldn't hurt to have a if (!tx.isFieldPresent(sfDelaySeconds)) return tefINTERNAL;`

Reply:
Although this is unlikely to happen in reality, I will add it as defensive check. (as `tecINTERNAL`)